fix ordinal numbers

fmatter opened the pull request:
English ordinals are complicated with their 1**st**, 2**nd**, 3**rd**, 4**th**. Weirdly enough, these ordinal suffixes are defined for 1-3 and 11-13 in the official CSL locales ([en-US](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-en-US.xml#L184-L191), [en-GB](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-en-GB.xml#L199-L206)). So the [`to_ordinal`](https://github.com/citeproc-py/citeproc-py/blob/d7fb0a0ff1a9858d15f83c5561e9846dce920699/citeproc/model.py#L1526-L1533) function fails miserably when it tries to generate an ordinal number larger than three, because it assumes `ordinal-04` to be defined. Changing the value to `ordinal-11` fixes the issue.

fmatter commented:
To demonstrate, see what happens if you increase the version number in the bib data:

```python
from citeproc import (Citation, CitationItem, CitationStylesBibliography,
                      CitationStylesStyle, formatter, source)

bib = source.json.CiteProcJSON(
    [
        {
            "type": "book",
            "id": "1",
            "title": "La Rettorica",
            "author": [{"literal": "Brunetto Latini"}],
            "issued": {"date-parts": [[1968]]},
            "editor": [{"literal": "Francesco Maggini"}],
            "edition": 3,
            "publisher": "Le Monnier",
            "location": "Firenze",
        }
    ]
)
bib_style = CitationStylesStyle("harvard1")
bibliography = CitationStylesBibliography(bib_style, bib, formatter.plain)
item = CitationItem("1")
bibliography.register(Citation([item]))
print(bibliography.style.render_bibliography([item])[0])
```

yarikoptic commented:
would you be so kind to add a test which would demonstrate the issue (fail) without the suggested fix?

fmatter commented:
Sure! I was not sure where to put it, so I just created a new file. The test fails with `tests/test_bibliography.py::TestBibliographyGeneration::test_generate - AttributeError: 'NoneType' object has no attribute 'single'`, because `get_term("ordinal-04")` returns `None` (while `get_term("ordinal-11")` returns an `<Element {http://purl.org/net/xbiblio/csl}term` with `.single == "th"`.

fmatter commented:
I was just reminded of this again because apparently I reinstalled citeproc-py in my venv.

Reading through it again, I really don't understand why the [CSL specification](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-en-US.xml#L184-L191) makes a jump from 4 to 11.
But since it also defines a default (`ordinal` without a suffix), I think it makes sense to use that instead of `ordinal-11`. 

It certainly doesn't make sense to ask for `ordinal-04`, which is never specified.

yarikoptic commented:
Your logic sounds reasonable to me. Frankly though I have not enough foo in citeproc/CSL to judge adequately.

FWIW ordinal-04 was originally  added by @brechtm  as `else` clause in e7c9f493ce2aed8033f597b50675dadce77a2f96 in 2011! And was added to CSL in https://github.com/citation-style-language/locales/commit/eb9d88e7aea99cbb336364b0b1a282d6d80efa89 in 2009 , and then removed in 2012 in https://github.com/citation-style-language/locales/commit/0374188ca0784c1d5cde3d1a2655897b9f4c8292 as to be replaced by ordinal (default - `th`), ordinal-11 -- ordinal-13. 
It even still exists in some locales:

```shell
 git describe --tags
v0.0.21
❯ git grep '"ordinal-04"'
locales-hi-IN.xml:    <term name="ordinal-04"></term>
locales-nl-NL.xml:    <term name="ordinal-04" match="last-two-digits">de</term>
locales-pa-PK.xml:    <term name="ordinal-04" gender-form="masculine">تھا</term>
locales-pa-PK.xml:    <term name="ordinal-04" gender-form="feminine">تھی</term>
❯ pwd
/home/yoh/proj/misc/citation-style-language--locales
```
So may be that would resolve at least part of the mystery ;)

zepinglee commented:
> Reading through it again, I really don't understand why the [CSL specification](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-en-US.xml#L184-L191) makes a jump from 4 to 11. But since it also defines a default (`ordinal` without a suffix), I think it makes sense to use that instead of `ordinal-11`.

The suffixes of "21st", "31st", "41st", ... match "1st" but "11th" is an exception in English. The terms `ordinal-04` to `ordinal-10` are missing because they can fall back to the default `ordinal` term "th" and so do 24–30, 34–40, 44–50, ...

zepinglee commented:
> FWIW ordinal-04 was originally added by @brechtm as `else` clause in [e7c9f49](https://github.com/citeproc-py/citeproc-py/commit/e7c9f493ce2aed8033f597b50675dadce77a2f96) in 2011! And was added to CSL in [citation-style-language/locales@eb9d88e](https://github.com/citation-style-language/locales/commit/eb9d88e7aea99cbb336364b0b1a282d6d80efa89) in 2009 , and then removed in 2012 in [citation-style-language/locales@0374188](https://github.com/citation-style-language/locales/commit/0374188ca0784c1d5cde3d1a2655897b9f4c8292) as to be replaced by ordinal (default - `th`), ordinal-11 -- ordinal-13. It even still exists in some locales:
> 
> ```shell
>  git describe --tags
> v0.0.21
> ❯ git grep '"ordinal-04"'
> locales-hi-IN.xml:    <term name="ordinal-04"></term>
> locales-nl-NL.xml:    <term name="ordinal-04" match="last-two-digits">de</term>
> locales-pa-PK.xml:    <term name="ordinal-04" gender-form="masculine">تھا</term>
> locales-pa-PK.xml:    <term name="ordinal-04" gender-form="feminine">تھی</term>
> ❯ pwd
> /home/yoh/proj/misc/citation-style-language--locales
> ```

This is a change from CSL v1.0 to v1.0.1 described in the [spec](https://docs.citationstyles.org/en/stable/specification.html#ordinal-suffixes).

> Ordinal terms work differently in CSL 1.0.1 and later than they did in CSL 1.0. When neither the style or locale file define the “ordinal” term, but do define the terms “ordinal-01” through “ordinal-04”, the original CSL 1.0 scheme is used: “ordinal-01” is used for numbers ending on a 1 (except those ending on 11), “ordinal-02” for those ending on a 2 (except those ending on 12), “ordinal-03” for those ending on a 3 (except those ending on 13) and “ordinal-04” for all other numbers.

fmatter commented:
Thank you for the feedback and the detective work :)
Looking at `to_ordinal()` again, it does not only fail for ordinals `> 3` in English, the entire function is also written **specifically for English** (or languages patterning like English I guess):

```python
def to_ordinal(number, context):
    number = str(number)
    last_digit = int(number[-1])
    if last_digit in (1, 2, 3) and not (len(number) > 1 and number[-2] == '1'):
        ordinal_term = 'ordinal-{:02}'.format(last_digit)
    else:
        ordinal_term = 'ordinal' # or 'ordinal-11' or 'ordinal-04'... 
    return number + context.get_term(ordinal_term).single
```

The long conditional describes numbers `X1`, `X2`, `X3` where `X!=11` -- i.e. **1st 2nd 3rd**, but not **11th 12th 13th**, but then again **21st 22nd 23rd** etc.
Of course, for other languages this makes no sense, for instance Dutch has **1ste, 2de, 3de** [...] **11de 12de 13de** [...] **21ste 22ste 23ste**; this means that `12` (*twaalfde*) and `13` pattern like `2` (*tweede*) and 3, but not like 22 (*tweeëntwintigste*) and 23, which is pretty much the opposite pattern from English.

Skimming through different locales and generating ordinals, I quickly realized that there was another problem: [if no ordinal term is found for a locale, it falls back to `en-US`](https://github.com/citeproc-py/citeproc-py/blob/d7fb0a0ff1a9858d15f83c5561e9846dce920699/citeproc/model.py#L98-L106).
This might make sense for other terms (? I have not looked at them), but it certainly doesn't for numbers.
To illustrate, [French](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-fr-FR.xml#L196-L199) has no definition for `ordinal-22` (because it's all **Xᵉ** except `1`); what should happen is that in its absence we fall back to `ordinal`, but what happens instead is that we get **22nd**!
(We also get **21ʳᵉ** because the English-centric function assumes that it should be like the ordinal for `1`, **1ʳᵉ**.
)
Long story short, I've rewritten `to_ordinal()` from scratch:

```python
def to_ordinal(number, context):
    if len(str(number)) == 1:
        ordinal_term = f'ordinal-{number:02}'
    else:
        ordinal_term = f'ordinal-{number}'
    if context.get_term(ordinal_term, fallback_locale=False) is None:
        ordinal_term = f'ordinal-{int(str(number)[-1]):02}'
        if context.get_term(ordinal_term, fallback_locale=False, zero_padded=True) is None:
            ordinal_term = f'ordinal'
    return str(number) + context.get_term(ordinal_term).single
```

It first pads single digits; if no term is found it tries a truncated and zero-padded last digit, e.g. `03` for `23`.
If this does not work, it falls back to `ordinal`.
I've added a new argument `fallback_locale=True` to `get_term()`, so we can specify that only the main locale should be used.
I've also added a new argument `zero_padded`, because for instance [Dutch](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-nl-NL.xml#L194-L213) specifies that e.g. `ordinal-05` should not match `25`, using the attribute `match='last-two-digits'` (instead of the default `'last-digit'`); `get_term()` needs to take that into account.
So now, this PR not only fixes the issue I originally set out to solve (English `> 3`), but also ordinals in most of the languages I looked at (the first column of a language is result of the current implementation of `to_ordinal`, the second column the result of my replacement suggestion):

en-US | [nl-NL](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-nl-NL.xml#L194-L213) |   | [es-ES](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-es-ES.xml#L193-L194) |   | [fr-FR](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-fr-FR.xml#L196-L199) |   | [de-DE](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-de-DE.xml#L205-L206) |   | [hi-IN](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-hi-IN.xml#L194-L200) |   | [it-IT](https://github.com/citation-style-language/locales/blob/e631a52dcea396be20d031b6456e91dba7772224/locales-it-IT.xml#L196-L199) |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1st | 1st | 1ste | 1st | 1.ª | 1ʳᵉ | 1ʳᵉ | 1st | 1. | 1 | 1 | 1st | 1º
2nd | 2de | 2de | 2nd | 2.ª | 2nd | 2ᵉ | 2nd | 2. | 2 | 2 | 2nd | 2º
3rd | 3de | 3de | 3rd | 3.ª | 3rd | 3ᵉ | 3rd | 3. | 3 | 3 | 3rd | 3º
4th | 4ste | 4de | 4.ª | 4.ª | 4ᵉ | 4ᵉ | 4. | 4. | 4वाँ | 4 | 4º | 4º
5th | 5ste | 5de | 5.ª | 5.ª | 5ᵉ | 5ᵉ | 5. | 5. | 5वाँ | 5वाँ | 5º | 5º
6th | 6ste | 6de | 6.ª | 6.ª | 6ᵉ | 6ᵉ | 6. | 6. | 6वाँ | 6 | 6º | 6º
7th | 7ste | 7de | 7.ª | 7.ª | 7ᵉ | 7ᵉ | 7. | 7. | 7वाँ | 7वाँ | 7º | 7º
8th | 8ste | 8ste | 8.ª | 8.ª | 8ᵉ | 8ᵉ | 8. | 8. | 8वाँ | 8वाँ | 8º | 8º
9th | 9ste | 9de | 9.ª | 9.ª | 9ᵉ | 9ᵉ | 9. | 9. | 9वाँ | 9वाँ | 9º | 9º
10th | 10ste | 10de | 10.ª | 10.ª | 10ᵉ | 10ᵉ | 10. | 10. | 10वाँ | 10वाँ | 10º | 10º
11th | 11ste | 11de | 11.ª | 11.ª | 11ᵉ | 11ᵉ | 11. | 11. | 11वाँ | 11 | 11º | 11º
12th | 12ste | 12de | 12.ª | 12.ª | 12ᵉ | 12ᵉ | 12. | 12. | 12वाँ | 12 | 12º | 12º
13th | 13ste | 13de | 13.ª | 13.ª | 13ᵉ | 13ᵉ | 13. | 13. | 13वाँ | 13 | 13º | 13º
14th | 14ste | 14de | 14.ª | 14.ª | 14ᵉ | 14ᵉ | 14. | 14. | 14वाँ | 14 | 14º | 14º
15th | 15ste | 15de | 15.ª | 15.ª | 15ᵉ | 15ᵉ | 15. | 15. | 15वाँ | 15वाँ | 15º | 15º
16th | 16ste | 16de | 16.ª | 16.ª | 16ᵉ | 16ᵉ | 16. | 16. | 16वाँ | 16 | 16º | 16º
17th | 17ste | 17de | 17.ª | 17.ª | 17ᵉ | 17ᵉ | 17. | 17. | 17वाँ | 17वाँ | 17º | 17º
18th | 18ste | 18de | 18.ª | 18.ª | 18ᵉ | 18ᵉ | 18. | 18. | 18वाँ | 18वाँ | 18º | 18º
19th | 19ste | 19de | 19.ª | 19.ª | 19ᵉ | 19ᵉ | 19. | 19. | 19वाँ | 19वाँ | 19º | 19º
20th | 20ste | 20ste | 20.ª | 20.ª | 20ᵉ | 20ᵉ | 20. | 20. | 20वाँ | 20वाँ | 20º | 20º
21st | 21st | 21ste | 21st | 21.ª | 21ʳᵉ | 21ᵉ | 21st | 21. | 21 | 21 | 21st | 21º
22nd | 22de | 22ste | 22nd | 22.ª | 22nd | 22ᵉ | 22nd | 22. | 22 | 22 | 22nd | 22º
23rd | 23de | 23ste | 23rd | 23.ª | 23rd | 23ᵉ | 23rd | 23. | 23 | 23 | 23rd | 23º
24th | 24ste | 24ste | 24.ª | 24.ª | 24ᵉ | 24ᵉ | 24. | 24. | 24वाँ | 24 | 24º | 24º
25th | 25ste | 25ste | 25.ª | 25.ª | 25ᵉ | 25ᵉ | 25. | 25. | 25वाँ | 25वाँ | 25º | 25º

Italian is still completely wrong because it should be **Xª** (*edizione* being feminine), but that's either a different issue or something to fix in the locale (note that Spanish gets it right by defining an "ungendered"/default `ordinal` **ª**).
Punjabi, which @yarikoptic mentioned is not included in the version CSL linked to citeproc-py, so yet another issue.
Finally, note that the orthographic convention for many languages not shown here is not to combine digits with suffixes, simply using `.` like German.

yarikoptic commented on citeproc/model.py:
```diff
@@ -95,14 +95,16 @@ def render(self, *args, **kwargs):
 
     # TODO: Locale methods
-    def get_term(self, name, form=None):
+    def get_term(self, name, form=None, fallback_locale=True, zero_padded=False):
```
better later (now) than never (before): with addition of these options and logic becoming more evolved, IMHO it is worth adding docstring describing them.

fmatter replied:
I agree.

fmatter commented:
How can I run the tests locally? I cannot use `pip install -e .` because it complains about a version number format.

fmatter commented:
I've removed `fallback_locale`, as I've realized that locales may underspecify terms and then have to fall back to a locale of the same language. Instead, `en-US` is now only added if there are no locales at all.

fmatter commented:
With a fresh 3.12 venv I get `packaging.version.InvalidVersion: Invalid version: 'a1f3e1f'`, also when using `tox`.
